src/target/adiv5.c

```diff
@@ -682,44 +683,72 @@ static void rp_rescue_setup(ADIv5_DP_t *dp)
 
 void adiv5_dp_init(ADIv5_DP_t *dp)
 {
-	/* designer code is in the same format in IDCODE and DPIDR */
-	/* common handling */
 	/*
-	 * the code in the DPIDR/IDCODE is in the form
+	 * Assume DP v1 or later.
+	 * this may not be true for JTAG-DP, in such cases DPIDR is not implemented
+	 * and reads are UNPREDICTABLE.
+	 *
+	 * for SWD-DP, we are guaranteed to be DP v1 or later.
+	 */
+	volatile uint32_t dpidr;
+	volatile struct exception e;
+	TRY_CATCH (e, EXCEPTION_ALL) {
+		dpidr = adiv5_dp_read(dp, ADIV5_DP_DPIDR);
+	}
+	if (e.type) {
+		DEBUG_WARN("DP not responding!...\n");
+		free(dp);
+		return;
+	}
+
+	/*
+	 * the code in the DPIDR is in the form
 	 * Bits 10:7 - JEP-106 Continuation code
 	 * Bits 6:0 - JEP-106 Identity code
 	 * here we convert it to our internal representation, See JEP-106 code list
-	 * note, this is the code of the designer not the implementer, we expect it to be ARM
+	 *
+	 * note: this is the code of the designer not the implementer, we expect it to be ARM
 	 */
-	const uint16_t designer = (dp->debug_port_id & ADIV5_DP_DPIDR_DESIGNER_MASK) >> ADIV5_DP_DPIDR_DESIGNER_OFFSET;
+	const uint16_t designer = (dpidr & ADIV5_DP_DPIDR_DESIGNER_MASK) >> ADIV5_DP_DPIDR_DESIGNER_OFFSET;
 	dp->designer_code =
 		(designer & ADIV5_DP_DESIGNER_JEP106_CONT_MASK) << 1U | (designer & ADIV5_DP_DESIGNER_JEP106_CODE_MASK);
+	dp->partno = (dpidr & ADIV5_DP_DPIDR_PARTNO_MASK) >> ADIV5_DP_DPIDR_PARTNO_OFFSET;
 
-	/* on JTAG IDCODE this is actually the version */
-	dp->revision = (dp->debug_port_id & ADIV5_DP_DPIDR_REVISION_MASK) >> ADIV5_DP_DPIDR_REVISION_OFFSET;
+	dp->mindp = !!(dpidr & ADIV5_DP_DPIDR_MINDP);
 
-	if (dp->version > 0) {
-		/* debug_port_id is valid DPIDR */
-		dp->partno = (dp->debug_port_id & ADIV5_DP_DPIDR_PARTNO_MASK) >> ADIV5_DP_DPIDR_PARTNO_OFFSET;
-		dp->mindp = !!(dp->debug_port_id & ADIV5_DP_DPIDR_MINDP);
-	} else {
-		/* called from JTAG scan, debug_port_id is JTAG TAP IDCODE */
-		dp->partno = (dp->debug_port_id & JTAG_IDCODE_PARTNO_MASK) >> JTAG_IDCODE_PARTNO_OFFSET;
-		dp->mindp = false;
-	}
+	dp->version = (dpidr & ADIV5_DP_DPIDR_VERSION_MASK) >> ADIV5_DP_DPIDR_VERSION_OFFSET;
```

Comment:
We should probably leave a note on how to do the JTAG DPv0 ID code stuff

Reply:
on dpv0 we dont really have an analogue, we just dont have DPIDR, unless we use the IDCODE, but that is also present on DPv1/2 so its not quite comparable. the handling of dpv0 is theres nothing to handle

Reply:
We're thinking along the lines that, we don't want to loose easy access to the version = 0 branch here for when we get to figuring out how to do DPv0